doc/BorderRouter.md

```diff
@@ -0,0 +1,118 @@
+Since a pool of packet buffers is bound to a receiver, the size of the pool can be configured for each
+receiver seperately.
+It makes sense to make this configurable on a per border router interface level because not every 
+interface might have the same expected load and hence not every interface requires the same pool size.
+An appropriate default value would have to be evaluated.
+
+## Number of processing routines (N)
+By configuring the number of processing routines one can specify the number of goroutines that are able
+to process packets in parallel.
+A default value could be 4 to 8 times the number of border router interfaces.
+
+## Number of slow-path processing routines (M)
+By configuring the number of slow-path processing routines one can specify the number of goroutines that
+process the packets on the slow-path.
+A default value could be the same number as we have border router interfaces.
+
+## Processing packets channel size
+Since each processing routine has a queue of packets to process and all packets not fitting in the queue
+are dropped, one has to specify a reasonable queue size.
+A default value could be 256.
+
+# Considerations for future work
+## Multiple Receivers per Border Router interface
+We could deploy multiple packet receivers per border router interface and use eBPF to make sure that all
+packets that belong to the same flow are received by the same receiver.
+Because the rest remains unchanged we would still have the "no-reordering" guarantee and significantly
+increase the read speed.
+
+## Lock goroutines to threads
+The CPU affiliation by locking the goroutines to threads and CPU cores can later be studied and would become
+much easier by applying this design because we have more goroutines and hence more flexibility how we want to fix
+them to the cores.
+
+## Replace go channels with custom ring buffer
+In the future we might want to replace the go channels that are used for communicating between the goroutines
+with custom ring buffers in case this provides higher performance.
+
+## Traffic engineering
+With the implementation as described in this document the forwarders process the packets in the order they are enqueued.
+In the future we can use additional queues for prioritized traffic between the processing routines and the forwarders.
+See [PR 4054](https://github.com/scionproto/scion/pull/4054).
+
+## UDP generic segment offloading (GSO)
+In the future we could add UDP generic segment offloading (GSO) for the connections between border router
+of different ASes to improve the performance even more.
+Such an implementation would be feasible in the future because we would just have to identify identify
+which border router interfaces are affected and for them make some changes to the IO parts.
+
+# Implementation steps
+* Restructure the router/dataplane.go file to have a reading, processing and forwarding functionality
+* Add buffer reusal support
+* Add slow-path support
```

Comment:
Add a step for the evaluation? Maybe the very first step even?
Ideally, the evaluation would also be part of the scion proto repo, e.g., a command that starts a very minimal topology, generates large traffic volumes, and measures the achieved performance and prints it to the terminal.

I am not sure whether it is indeed needed to push it to scion proto, but I guess it would anyway make sense for you to implement the performance evaluation in a clean and reusable way, even in case it remains on one of your own branches.

Reply:
The question would be how to measure the performance. For the COLIBRI Gateway I repeatedly queried the metrics endpoint but this would require a custom implementation or at least a modification of an already existing tool. Do you already have a tool that does something like that, maybe even with terminal support?

Reply:
I don't think there is a tool for that already.
How much would you need to change in the BR to do something similar to what you did with the COLIBRI gateway?
There are already metrics exposed in the BR, you will probably have to adapt them anyway a bit with the new proposal.
Maybe you can also use this approach to measure the performance?

For the traffic generation, maybe the pktgen tool could work?
https://github.com/scionproto/scion/blob/master/tools/pktgen/cmd/pktgen/main.go#L182-L183

Reply:
In the border router not much. We just need a tool to query the metrics. For the COLIBRI Gateway that querying was already included in the traffic generator tool. I could create a tool that just queries the metrics endpoint. This should not take much time.